### DISCUSSION TITLE
cargo outdated error

### DISCUSSION THREAD

**Discussion by tessus** [dani-garcia/vaultwarden#4324] [Q&A] [answered]
Getting the following when running `cargo outdated -R` in the vaultwarden project root:

```
$ cargo outdated -R
error: failed to select a version for `libsqlite3-sys`.
    ... required by package `diesel v2.1.4`
    ... which satisfies dependency `diesel = "^2.1"` of package `diesel_logger v0.3.0`
    ... which satisfies dependency `diesel_logger = "^0.3.0"` of package `vaultwarden v1.0.0 (/var/folders/3d/576s_9xd0fn73g9zn9f6cxf40000gn/T/cargo-outdatedathJqz)`
versions that meet the requirements `>=0.17.2, <0.28.0` are: 0.27.0, 0.26.0, 0.25.2, 0.25.1, 0.25.0, 0.24.2, 0.24.1, 0.24.0, 0.23.2, 0.23.1, 0.23.0, 0.22.2, 0.22.1, 0.22.0, 0.20.1, 0.20.0, 0.18.0, 0.17.3, 0.17.2

the package `libsqlite3-sys` links to the native library `sqlite3`, but it conflicts with a previous package which links to `sqlite3` as well:
package `libsqlite3-sys v0.28.0`
    ... which satisfies dependency `libsqlite3-sys = "^0.28.0"` of package `vaultwarden v1.0.0 (/var/folders/3d/576s_9xd0fn73g9zn9f6cxf40000gn/T/cargo-outdatedathJqz)`
Only one package in the dependency graph may specify the same links value. This helps ensure that only one copy of a native library is linked in the final binary. Try to adjust your dependencies so that only one package uses the links ='libsqlite3-sys' value. For more information, see https://doc.rust-lang.org/cargo/reference/resolver.html#links.

failed to select a version for `libsqlite3-sys` which could resolve this conflict
```

**Comment by stefan0xC** [comment, marked as answer]
Not sure if this is a vaultwarden issue. You can exclude the package by calling `cargo outdated -x libsqlite3-sys`

**Reply by BlackDex** [reply]
That is what i do too.

**Reply by tessus** [reply]
interesting. thanks for the info. it's just that I haven't seen this before. The last time I ran this command, everything was good and I didn't get that error.

**Reply by BlackDex** [reply]
The problem is somewhere in cargo-outdated. It seems to be unable to uphold the version constraints mandated by diesel.

**Reply by stefan0xC** [reply]
Should be temporarily resolved by the next release of diesel https://github.com/diesel-rs/diesel/commit/c4c0c8ac11bdc53e95f45d1ce91560ad442a6baf  (unless a `libsqlite3-sys 0.29.0` is released before then.) The underlying issue seems to be https://github.com/kbknapp/cargo-outdated/issues/376 (i.e. `cargo outdated` not taking into account version constraints when trying to resolve dependencies.)

edit: Just tested it and the error message seems to be the same as when I change the version of  https://github.com/dani-garcia/vaultwarden/blob/897bdf8343148ee4e4f1a937cbc501ad3f743b4a/Cargo.toml#L83 manually to `0.28.0` and run `cargo update`.

**Reply by BlackDex** [reply]
That is because diesel doesn't support that version. Maybe it use the way diesel sets the version constraint?

**Reply by stefan0xC** [reply]
I think it's generally a problem if future versions are incompatible with existing version constraints. According to https://github.com/kbknapp/cargo-outdated/pull/271 the current solution is to exclude the package.
